src/controller/ArchipelagoController.ts

```diff
@@ -0,0 +1,212 @@
+      const { requestId, error } = message
+      this.activeWorkerRequests[requestId]?.reject(error)
+      delete this.activeWorkerRequests[requestId]
+    } else if ("requestId" in message) {
+      const { requestId, payload } = message as WorkerResponse
+      this.activeWorkerRequests[requestId]?.resolve(payload)
+      delete this.activeWorkerRequests[requestId]
+    }
+  }
+
+  sendMessageToWorker(message: WorkerMessage) {
+    this.worker.send(message)
+  }
+
+  sendRequestToWorker<T>(message: Omit<WorkerRequest, "requestId">) {
+    const requestId = this.requestIdGenerator.generateId()
+
+    return new Promise<T>((resolve, reject) => {
+      this.activeWorkerRequests[requestId] = { resolve, reject }
+
+      this.sendMessageToWorker({ ...message, requestId } as WorkerMessage)
+
+      setTimeout(() => {
+        if (this.activeWorkerRequests[requestId]) {
+          delete this.activeWorkerRequests[requestId]
+          reject(new Error("Request timed out"))
+        }
+      }, this.options.requestTimeoutMs)
+    })
+  }
+
+  async dispose() {
+    await this.sendRequestToWorker({ type: "dispose-request" })
+    this.worker.kill()
+  }
+}
+
+export class ArchipelagoControllerImpl implements ArchipelagoController {
+  pendingUpdates: Map<string, PeerUpdate> = new Map()
+
+  updatesSubscribers: Set<UpdateSubscriber> = new Set()
+
+  activePeers: Map<string, PeerData> = new Map()
+  flushFrequency: number
+  logger: Logger
+
+  workerController: WorkerController
+
+  disposed: boolean = false
+
+  constructor(options: ArchipelagoControllerOptions) {
+    this.flushFrequency = options.flushFrequency ?? 2
+    this.logger = options.logger ?? console
+    this.workerController = new WorkerController(this.handleWorkerMessage.bind(this), options.archipelagoParameters)
+
+    this.startFlushLoop()
+  }
+
+  startFlushLoop() {
+    const loop = () => {
+      if (!this.disposed) {
+        this.flush()
+        setTimeout(loop, this.flushFrequency * 1000)
+      }
+    }
+
+    loop()
+  }
+
+  async flush() {
+    if (this.pendingUpdates.size > 0 && this.workerController.workerStatus === "idle") {
+      console.log(`Flushing ${this.pendingUpdates.size} updates`)
+      const updatesToFlush = this.pendingUpdates
+      this.pendingUpdates = new Map()
+
+      const positionUpdates: PeerPositionChange[] = []
+      const clearUpdates: string[] = []
+
+      for (const [id, update] of updatesToFlush) {
+        if (update.type === "set-position") {
+          positionUpdates.push({ id, position: update.position })
+        } else {
+          clearUpdates.push(id)
+        }
+      }
+
+      this.workerController.sendMessageToWorker({ type: "apply-updates", updates: { positionUpdates, clearUpdates } })
+    }
+  }
+
+  setPeersPositions(...requests: PeerPositionChange[]): void {
+    for (const req of requests) {
+      if (!this.activePeers.has(req.id)) {
+        this.activePeers.set(req.id, { id: req.id, position: req.position })
```

Comment:
I don't think we are using the PeerData at all. Would it make sense to have `activePeers` be a `Set` of ids instead?

Reply:
Here is a contentious point. I was thinking that we may want to have the information "cached" in the controller. But I'm not sure. I think I will do as you say here. Having redundant information is never a good time.

Reply:
Fun fact: Active peers were a Set in the implementation that I did in the other branch.

Reply:
Maybe it can be a set and we can change it in the future if we need to?